Dockerfile

```diff
@@ -37,6 +38,8 @@ ENV DJANGO_SETTINGS_MODULE=normandy.settings \
     NEW_RELIC_CONFIG_FILE=newrelic.ini
 EXPOSE $PORT
 
+ENTRYPOINT ["/bin/bash", "/app/ci/run.sh"]
+
 CMD $CMD_PREFIX \
```

Comment:
I'm a little nervous about having both `ENTRYPOINT` and `CMD`, but after reading some docs and looking closely at `run.sh`, I think it will be fine. It might be nice to have a comment here or in `run.sh`'s `*)` case explaining whats going on. I'm also not 100% sure the `$CMD_PREFIX` stuff will work.

@sciurus can you remind me what we're using `$CMD_PREFIX` for? and do you think that this along with the `*)` case in `run.sh` will work out fine with it?

Reply:
This is a little convoluted but it should work.

everything is started as an argument to run.sh. For a few arguments (e.g. `lint`) it has a predefined set of commands to run. Anything it doesn't recognize it will just pass through to the shell.

CMD_PREFIX is undefined except for on the APM node. There it is defined as new relic's wrapper script.

Reply:
I'm mostly concerned about the situation in which `CMD_PREFIX` is not undefined. Will `exec "$@"` work fine in that case? I guess it depends on what `CMD_PREFIX` is set to?

Reply:
PS: I'd still like to see a comment, probably in both the `Dockerfile` and `run.sh` about what's going on. It may also make sense to move run.sh out of `ci/`. Maybe keeping it all in `bin/` is a better idea?

Reply:
Regarding when CMD_PREFIX is defined, sure, it should work if that's a valid command. In our case it would be the difference between `exec gunicorn` and `exec newrelic-admin run-program gunicorn`

Reply:
@peterbe the dockerfile is just started via a systemd unit file which declares

```
ExecStart=/usr/bin/docker run \
        --net host -m 0b --env-file /etc/dockerflow/app.txt -v /usr/share/GeoIP/GeoIP2-Country.mmdb:/app/GeoIP2-Country.mmdb --log-driver=none  \
        --name app \
        hub.prod.mozaws.net/pipelines/normandy:v97-422 \
```

for the current version. So you can see it runs the default `CMD`.

I am thinking in a similar vein to you. It would be simpler if anything we anticipated running was defined in `run.sh`, and `CMD` just specified one of those to make it the default. In `run.sh` we could have

``
web)
$CMD_PREFIX \
    gunicorn  \
    --log-file - \
    --worker-class ${GUNICORN_WORKER_CLASS:-sync} \
    --max-requests ${GUNICORN_MAX_REQUESTS:-0} \
normandy.wsgi:application
;;
```

and in the Dockerfile `CMD` could change to

`CMD web`

We could bikeshed the name `web`.

Reply:
There's another subtle refactoring. The `Dockerfile` does `yarn build` and `manage.py collectstatic` also. Which is repeated in `ci/run.sh`. 

I was in love with the idea that the `ci/` directory was "self contained" and was only used for CI and if you one day decide to not run CI anymore we could just delete this directory and nothing would change on the side of running things. 

Perhaps we `git mv ci/run.sh bin/run.sh` (and fix other relative paths accordingly) and make that `run.sh` the home for all bash logic such as `collectstatic`, karma tests, pytests etc. 

We *could* add a couple of entries inside that `bin/run.sh` that calls `ci/run.sh` instead. 

In conclusion; 
* less bash in `Dockerfile`, move it to `run.sh`
* use the current `ci/run.sh` for other non-CI related things such as "web"
* change the `CMD` line to be just `CMD web`.

By the way, this is pretty much how [Tecken does it](https://github.com/mozilla-services/tecken/blob/786b1da12a8163d66a6e86eb6a258661b29b34b1/Dockerfile#L72-L74).

Reply:
So @mythmon @sciurus Do you like that plan? I.e. remove the extra bash from the `Dockerfile` and move it into `bin/run.sh` and edit `bin/run.sh` so that it can execute the other `docker-compose` commands we run in CI.

Reply:
👍

Reply:
I think this sounds good.